agent/acs/handler/payload_handler.go

```diff
@@ -187,14 +184,24 @@ func (payloadHandler *payloadRequestHandler) handleSingleMessage(payload *ecsacs
 	go func() {
 		// Throw the ack in async; it doesn't really matter all that much and this is blocking handling more tasks.
 		for _, credentialsAck := range credentialsAcks {
-			payloadHandler.refreshHandler.ackMessage(credentialsAck)
+			payloadHandler.makeCredentialsAckRequest(credentialsAck)
 		}
 		payloadHandler.ackRequest <- *payload.MessageId
 	}()
 
 	return nil
 }
 
+// makeCredentialsAckRequest sends an IAMRoleCredentialsAckRequest to the backend
+func (payloadHandler *payloadRequestHandler) makeCredentialsAckRequest(ack *ecsacs.IAMRoleCredentialsAckRequest) {
+	seelog.Debugf("ACKing credentials: %s", ack.String())
+	err := payloadHandler.acsClient.MakeRequest(ack)
+	if err != nil {
+		seelog.Warnf("Error ACKing credentials with credentialsID '%s', error: %v",
```

Comment:
Q: what's the impact of failing to ACK credentials? Asking because I am not sure if warn is appropriate level.  (nvm I see `// Throw the ack in async; it doesn't really matter all that much and this is blocking handling more tasks.`)

I also see in ecs-agent/../refresh_credentials_responder we have very similar logging and both does ack. Can you explain the difference between these two? I might be missing something but we might be able to get rid of one of the two?

Reply:
Thanks for taking a look!

> I also see in ecs-agent/../refresh_credentials_responder we have very similar logging and both does ack. Can you explain the difference between these two? I might be missing something but we might be able to get rid of one of the two?

Payload responder receives payload messages which include tasks and their associated credentials. If the payload message is successfully processed, the credentials associated with the tasks in that message are ACKed.

Refresh credentials responder receives refresh credentials messages to refresh credentials that may be a risk of expiring. If the credentials are successfully refreshed, the credentials associated with the message is ACKed.

I think it would be better to have two different functions/logging for these 2 different cases/code paths mentioned above, that way in the logs we can easily see whether the ACK came from handling a payload message or handling a refresh credentials message. In a future PR to refactor ACS payload message handling, I plan to update the logging message there to make it explicitly clear that the ACKing of credentials being done there is associated with credentials from a payload message. If you have a strong opinion/preference, I can pull that into the scope of this pull request. What do you think?

Reply:
This makes sense to me. One question I had is that if we are getting a refresh credentials message from ACS, would the message be handled by both of `payloadHandler` and `refreshCredentialsHandler`? If so, that would cause us to call `setTaskCredentials` (may not exactly the method name) twice. What would that cause us? If one of the Acks failed, what should we expect? If not, how did we assure it only gets to refreshCredentialsHandler?

Reply:
Synced up and discussed offline:
- Each ACS message has its own type, which each message responder is specialized to handle
- Refresh credentials messages (type IAMRoleCredentialsMessage) received from ACS would only be handled by the refresh credentials responder
- Payload responder handles payload messages (type PayloadMessage) which includes handling the credentials that are associated with task(s) contained in the payload message

We aligned that we should make an update in this PR to to make logging of ACKing credentials while handling  payload message explicitly distinct from ACKing credentials while handling refresh credentials message.